zcash_primitives/src/zip32.rs

```diff
@@ -43,41 +48,54 @@ impl ConditionallySelectable for AccountId {
 
 // ZIP 32 structures
 
-/// A child index for a derived key
+/// A child index for a derived key.
+///
+/// Only hardened derivation is supported.
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub enum ChildIndex {
-    NonHardened(u32),
-    Hardened(u32), // Hardened(n) == n + (1 << 31) == n' in path notation
-}
+pub struct ChildIndex(u32);
 
 impl ChildIndex {
-    pub fn from_index(i: u32) -> Self {
-        match i {
-            n if n >= (1 << 31) => ChildIndex::Hardened(n - (1 << 31)),
-            n => ChildIndex::NonHardened(n),
+    /// Parses the given ZIP 32 child index.
+    ///
+    /// Returns `None` if the hardened bit is not set.
+    pub fn from_index(i: u32) -> Option<Self> {
```

Comment:
I would almost prefer that this operation be called `decode` or something of the sort, pursuant to my previous comment.